app/models/term.rb

```diff
@@ -0,0 +1,150 @@
+      'uri'           => uri,
+      'pref_label'    => pref_label,
+      'alt_labels'    => alt_labels,
+      'term_type'     => term_type,
+      'vocabulary'    => vocabulary.string_key,
+      'authority'     => authority,
+      'custom_fields' => custom_fields.to_json
+    }.tap do |doc|
+      vocabulary.custom_fields.each do |k, v|
+        doc["#{k}#{Solr::Utils.suffix(v[:data_type])}"] = custom_fields[k]
+      end
+    end
+  end
+
+  def set_custom_field(field, value)
+    custom_fields[field] = value
+  end
+
+  def self.local_uri_prefix
+    host = HYACINTH[:local_uri_prefix]
+
+    raise 'Missing local_uri_prefix in config/hyacinth.yml' unless host
+    host.ends_with?('/') ? host : "#{host}/"
+  end
+
+  private
+
+    def cast_custom_fields
+      custom_fields.each do |k, v|
+        next if v.nil?
+        raise "custom_field #{k} is not a valid custom field" unless vocabulary.custom_fields.keys.include?(k)
+
+        data_type = vocabulary.custom_fields[k][:data_type]
+
+        # check data type valid
+        raise "custom_field #{k} must be a valid #{data_type}. Could not cast \"#{v}\" to a valid #{data_type}" unless send("valid_#{data_type}?", v)
+
+        # skip, if not string value
+        next unless v.is_a?(String)
+
+        # cast string values
+        case data_type
+        when 'integer'
+          custom_fields[k] = v.to_i
+        when 'boolean'
+          custom_fields[k] = v == 'true'
+        end
+      end
+    end
+
+    def validate_custom_fields
+      custom_fields.each do |k, v|
+        next if v.nil?
+
+        if vocabulary.custom_fields.keys.include?(k)
+          data_type = vocabulary.custom_fields[k][:data_type]
+          errors.add(:custom_field, "#{k} must be a valid #{data_type}") unless send("valid_#{data_type}?", v)
+        else
+          errors.add(:custom_field, "#{k} is not a valid custom field.")
+        end
+      end
+    end
+
+    def valid_string?(v)
+      v.is_a?(String)
+    end
+
+    def valid_integer?(v)
+      v.is_a?(Integer) || (v.is_a?(String) && /\A[+-]?[1-9]\d*\z/.match(v))
+    end
+
+    def valid_boolean?(v)
+      (!!v == v) || (v.is_a?(String) && /\A(true|false)\z/.match(v))
+    end
+
+    def set_uid
+      raise StandardError, 'Cannot set uid if record has already been persisted.' unless new_record?
+
+      self.uid = SecureRandom.uuid unless uid
+    end
+
+    def set_uri
+      case term_type
+      when LOCAL
+        self.uri = "#{Term.local_uri_prefix}term/#{uid}"
+      when TEMPORARY
+        self.uri = URI(TEMPORARY_URI_BASE + Digest::SHA256.hexdigest(vocabulary.string_key + pref_label)).to_s
+      end
+    end
+
+    def set_uri_hash
+      self.uri_hash = Digest::SHA256.hexdigest(uri) if uri
+    end
+
+    # Check that uid, uri and term_type were not changed.
+    def uid_uri_and_term_type_unchanged
+      return unless persisted? # skip if object is new or is deleted
+
+      errors.add(:uid, 'Change of uid not allowed!') if uid_changed?
```

Comment:
Compared to the old URI service gem, it's nice extending ApplicationRecord and getting the *_changed? methods automatically.